RATIONALE.md

```diff
@@ -0,0 +1,61 @@
+# Notable rationale of the WebAssembly ABI
+
+## Why is score i32 not i64?
+
+`framework.ScorePlugin` currently declares score as int64. In practice, this
+seems an accident. For example, negative scores don't have semantic meaning
+defined in docs. Also, there is no known use case that requires more than 32
+bit range. For example, even if signed, justifying a 64-bit score implies two
+million values is not broad enough. In summary, score seems both accidentally
+signed and accidentally 64-bit.
+
+So, there still remains a question why not leave score as 64-bit anyway?
+
+The reason is that most compilers target WebAssembly Core Specification 1.0,
+the only REC status specification in w3c. This allows up to one result of up to
+64bits.
+
+We currently use 32-bits to return the status as an i32. It is still easy to
+return a second 32-bit value because two can be packed into a single i64.
+
+When the combined results require more than 64 bits, we have to share memory to
+return more results, as out pointer parameters. This is complicated when the
+function is called host->guest, because the guest controls memory. For the host
+to pass an out-pointer, it needs a region of memory unused by the guest.
+
+Most often, this is reserved with exported functions. For example, TinyGo
```

Comment:
inspired by @ncruces approach with sqlite (reserve memory for whole module use) we could also make a single init function which the host passes how much scratch space. the guest keeps an active pointer to this the whole life of the module. Free isn't needed because it is reserved the entire time. This would be toolchain independent and work for many fixed width things like score, but not for things like name which are unbounded.

Reply:
I think for score alone, this is overkill, but anyway food for thought in case the topic comes up again.